pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,951 @@
+			peerToUpdateKeys.Insert(peerKey)
+		}
+	}
+	peerToDeleteKeys := prePeerKeys.Difference(curPeerKeys)
+
+	bgpServer := c.bgpPolicyState.bgpServer
+	for key := range peerToAddKeys {
+		peerConfig := curPeerConfigs[key]
+		if err := bgpServer.AddPeer(c.ctx, peerConfig); err != nil {
+			return err
+		}
+		c.bgpPolicyState.peerConfigs[key] = peerConfig
+	}
+	for key := range peerToUpdateKeys {
+		peerConfig := curPeerConfigs[key]
+		if err := bgpServer.UpdatePeer(c.ctx, peerConfig); err != nil {
+			return err
+		}
+		c.bgpPolicyState.peerConfigs[key] = peerConfig
+	}
+	for key := range peerToDeleteKeys {
+		peerConfig := prePeerConfigs[key]
+		if err := bgpServer.RemovePeer(c.ctx, peerConfig); err != nil {
+			return err
+		}
+		delete(c.bgpPolicyState.peerConfigs, key)
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcileBGPAdvertisements(bgpAdvertisements v1alpha1.Advertisements) error {
+	curRoutes, err := c.getRoutes(bgpAdvertisements)
+	if err != nil {
+		return err
+	}
+	preRoutes := c.bgpPolicyState.routes
+	routesToAdvertise := curRoutes.Difference(preRoutes)
+	routesToWithdraw := preRoutes.Difference(curRoutes)
+
+	bgpServer := c.bgpPolicyState.bgpServer
+	for route := range routesToAdvertise {
+		if err := bgpServer.AdvertiseRoutes(c.ctx, []bgp.Route{route}); err != nil {
+			return err
+		}
+		c.bgpPolicyState.routes.Insert(route)
+	}
+	for route := range routesToWithdraw {
+		if err := bgpServer.WithdrawRoutes(c.ctx, []bgp.Route{route}); err != nil {
+			return err
+		}
+		c.bgpPolicyState.routes.Delete(route)
+	}
+
+	return nil
+}
+
+func hashNodeNameToIP(s string) string {
+	h := fnv.New32a() // Create a new FNV hash
+	h.Write([]byte(s))
+	hashValue := h.Sum32() // Get the 32-bit hash
+
+	// Convert the hash to a 4-byte slice
+	ip := make(net.IP, 4)
+	ip[0] = byte(hashValue >> 24)
+	ip[1] = byte(hashValue >> 16)
+	ip[2] = byte(hashValue >> 8)
+	ip[3] = byte(hashValue)
+
+	return ip.String()
+}
+
+func (c *Controller) getRouterID() (string, error) {
+	// According to RFC 4271:
+	// BGP Identifier:
+	//   This 4-octet unsigned integer indicates the BGP Identifier of
+	//   the sender.  A given BGP speaker sets the value of its BGP
+	//   Identifier to an IP address that is assigned to that BGP
+	//   speaker.  The value of the BGP Identifier is determined upon
+	//   startup and is the same for every local interface and BGP peer.
+	//
+	// In goBGP, only an IPv4 address can be used as the BGP Identifier (BGP router ID).
+	// For IPv4-only or dual-stack Kubernetes clusters, the Node's IPv4 address is used as the BGP router ID, ensuring
+	// uniqueness.
+	// For IPv6-only Kubernetes clusters without a Node IPv4 address, the router ID could be specified in the Node
+	// annotation `node.antrea.io/bgp-router-id`. If the annotation is not present, an IPv4 address will be generated by
+	// hashing the Node name and updated to the Node annotation `node.antrea.io/bgp-router-id`. If the annotation is
```

Comment:
Remove "If the annotation is present...". It is already covered by the first sentence.